esp/esdllib/esdl_xpath_extensions_libxml.cpp

```diff
@@ -472,6 +474,326 @@ static void strTokenizeFunction(xmlXPathParserContextPtr ctxt, int nargs)
+            {
+                xmlFree(tmp);
+                xmlXPathSetError((ctxt), XPATH_EXPR_ERROR);
+                return;
+            }
+        }
+        xmlFree(tmp);
+    }
+    if (2 <= nargs)
+    {
+        tmp = xmlXPathPopString(ctxt);
+        if (xmlXPathCheckError(ctxt))
+            return;
+        if (*tmp)
+        {
+            if (strieq((const char*)tmp, "normal") || streq((const char*)tmp, "5"))
+                level = LogNormal;
+            else if (strieq((const char*)tmp, "max") || streq((const char*)tmp, "10"))
+                level = LogMax;
+            else if (!strieq((const char*)tmp, "min"))
+            {
+                if (TokenDeserializer().deserialize((const char*)tmp, level) != Deserialization_SUCCESS || level < LogMin || level > LogMax)
+                {
+                    xmlFree(tmp);
+                    xmlXPathSetError((ctxt), XPATH_EXPR_ERROR);
+                    return;
+                }
+            }
+        }
+        xmlFree(tmp);
+    }
+    if (1 <= nargs)
+    {
+        tmp = xmlXPathPopString(ctxt);
+        if (xmlXPathCheckError(ctxt))
+            return;
+        if (*tmp)
+        {
+            if (strieq((const char*)tmp, "json"))
+                style = TXSUMMARY_OUT_JSON;
+            else if (!strieq((const char*)tmp, "text"))
+            {
+                xmlFree(tmp);
+                xmlXPathSetError((ctxt), XPATH_EXPR_ERROR);
+                return;
+            }
+        }
+        xmlFree(tmp);
+    }
+
+    StringBuffer output;
+    txSummary->serialize(output, level, group, style);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar*)output.str()));
+}
+
+static void deprecatedEncryptString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toEncrypt = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer encrypted;
+    try
+    {
+        encrypt(encrypted, (const char*)toEncrypt);
+    }
+    catch (IException* e)
+    {
+        e->Release();
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    catch (...)
+    {
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    if (!xmlXPathCheckError(ctxt))
+        xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encrypted.str()));
+    xmlFree(toEncrypt);
+}
+
+static void deprecatedDecryptString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecrypt = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decrypted;
+    try
+    {
+        decrypt(decrypted, (const char*)toDecrypt);
+    }
+    catch (IException* e)
+    {
+        e->Release();
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
```

Comment:
Not sure this would be an operand error.  Perhaps just go with the vague XPATH_UNKNOWN_FUNC_ERROR?

Reply:
My thought was that "invalid operand" was closest to "invalid parameter", which seems applicable since the parameter value is not valid for the intended use. "Unknown function" sounds more like the function name itself is invalid, instead of the parameters passed, which is more confusing to me than vague.

Reply:
Looking at the enumeration again, perhaps `XPATH_EXPR_ERROR` provides an appropriate level of vagueness if it just means something in XPath is wrong.

Reply:
I see what you mean about the likely cause being the content of the parameter.. but we don't really know that.

But there are no great choices.  They made explicit errors for their needs and didn't leave anything similar to an unknown error.

In libxml2 the cases of XPATH_INVALID_OPERAND error were because the operand couldn't be retrieved.

We could consider using XPTR_RESOURCE_ERROR as a vague error.  I don't even see it used in libxml2.  That would be thinking of our extension as a resource I guess.  But it's also a stretch.  So no good choices basically means it doesn't matter as long as we don't imply something way off that libxml2 may act on.

Reply:
Sure XPATH_EXPR_ERROR would be fine.